### PR TITLE
Expose ExchangeSource metrics in query json

### PR DESCRIPTION
**Pull request by losipiuk**
## Description

Expose ExchangeSource metrics in query json

Another approach after reverting https://github.com/trinodb/trino/pull/22685.

This time we introduce pipelineOperatorMetrics with explicit property of being shared between all the operator instances for given operator id within a pipeline.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/PipelineContext.java [hunk 1]
```diff
@@ -208,7 +216,8 @@ public void driverFinished(DriverContext driverContext)
         // merge the operator stats into the operator summary
         List<OperatorStats> operators = driverStats.getOperatorStats();
         for (OperatorStats operator : operators) {
-            operatorSummaries.merge(operator.getOperatorId(), operator, OperatorStats::add);
+            Metrics pipelineLevelMetrics = pipelineOperatorMetrics.getOrDefault(operator.getOperatorId(), Metrics.EMPTY);
+            operatorSummaries.merge(operator.getOperatorId(), operator, (stats1, stats2) -> stats1.addFillingPipelineMetrics(stats2, pipelineLevelMetrics));
```
Comment (wendigo) [inline core/trino-main/src/main/java/io/trino/operator/PipelineContext.java:220]: nit: s/stats1/first/
s/stats2/second ?

File: core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSource.java [hunk 2]
```diff
@@ -90,6 +92,15 @@ public interface ExchangeSource
      */
     long getMemoryUsage();
 
+    /**
+     * Get metrics describing this ExchangeSource.
```
Comment (wendigo) [inline core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSource.java:96]: drop 'this'

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java [hunk 3]
```diff
@@ -296,11 +300,14 @@ public AzureExchangeStorageReader(
                 BlobServiceAsyncClient blobServiceAsyncClient,
                 List<ExchangeSourceFile> sourceFiles,
                 int blockSize,
-                int maxPageStorageSize)
+                int maxPageStorageSize,
```
Comment (wendigo) [inline plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java:303]: move `metricsBuilder` below `sourceFiles` ? I prefer to keep objects together and primitive values separately and do not mix one with each other

**Review by wendigo — APPROVED** [review]

File: plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java (hunk 3 above)
Comment (losipiuk) [reply plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/azure/AzureBlobFileSystemExchangeStorage.java]: I do not think it is generally beneficial - I would rather try to group logically. But I do not care in this case so moved.

**Pull request merged**